File: src/cmidas/parser.c
#include <stdbool.h>
#include <stdlib.h>

#include "error.h"
#include "lexer.h"
#include "parser.h"
#include "vector.h"

         Vector_T  program    (struct par_state *);
static struct stm *declaration(struct par_state *);
static struct stm *statement  (struct par_state *);
static struct stm *block      (struct par_state *);
static struct stm *if_cond    (struct par_state *);
static struct stm *while_cond (struct par_state *);
static struct stm *var_decl   (struct par_state *);
static struct stm *expr_stmt  (struct par_state *);
static struct stm *print      (struct par_state *);

static struct exp *expression    (struct par_state *);
static struct exp *assignment    (struct par_state *);
static struct exp *equality      (struct par_state *);
static struct exp *ordering      (struct par_state *);
static struct exp *addition      (struct par_state *);
static struct exp *multiplication(struct par_state *);
static struct exp *unary         (struct par_state *);
static struct exp *primary       (struct par_state *);

static struct tok *tok_next   (struct par_state *);
       static bool tok_matches(struct par_state *, enum tok_type);
inline static bool tok_is     (struct par_state *, enum tok_type);
inline static bool tok_was    (struct par_state *, enum tok_type);

static struct stm *stm_new_block    (Vector_T);
static struct stm *stm_new_if       (struct exp *, struct stm *, struct stm *);
static struct stm *stm_new_while    (struct exp *, struct stm *);
static struct stm *stm_new_var_decl (struct tok *, struct exp *);
static struct stm *stm_new_print    (struct exp *, struct tok *);
static struct stm *stm_new_expr_stmt(struct exp *, struct tok *);

static struct exp *exp_new_assign (struct tok *, struct exp *);
static struct exp *exp_new_binary (struct tok *, struct exp *, struct exp *);
static struct exp *exp_new_unary  (struct tok *, struct exp *);
static struct exp *exp_new_group  (struct exp *, struct tok *, struct tok *);
static struct exp *exp_new_var    (struct tok *);
static struct exp *exp_new_literal(struct tok *);

Vector_T
parse(struct par_state *par, const char *path)
{
    lex_init(&par->lex);
    par->path = path;
    par->prev_tok = NULL;
    par->this_tok = NULL;

    lex_feed(&par->lex, path);
    tok_next(par);

    return program(par);
}

Vector_T
program(struct par_state *par)
/*  program -> statement* "EOF"                                              */
{
    Vector_T program;

    program = Vector_new(sizeof(struct stm *));
    /* XXX if we would have used !tok_matches(par, TOK_EOF), we would force the
       lexer to read beyond the EOF upon finally matching it */
    while(!tok_is(par, TOK_EOF))
    {
        struct stm *stm = declaration(par);
        Vector_push(program, &stm);
    }

    return program;
}

static struct stm *
declaration(struct par_state *par)
/*  declaration -> var_decl ";"
                 | statement                                                 */
{
    struct stm *stm;

    if (tok_matches(par, TOK_VAR))
        stm = var_decl(par);
    else
        stm = statement(par);

    return stm;
}

static struct stm *
var_decl(struct par_state *par)
/*  var_decl -> ^var^ identifier "=" expression ";"                          */
{
    struct stm *stm;

    if (tok_matches(par, TOK_IDENTIFIER))
    {
        struct tok *name = par->prev_tok;
        if (tok_matches(par, TOK_EQUAL))
        {
            struct exp *value = expression(par);
            stm = stm_new_var_decl(name, value);
        }
        else
            err_at_tok(par->path, par->prev_tok,
                "\n    An equal sign should follow the variable's name.\n\n");
    }
    else
        err_at_tok(par->path, par->prev_tok,
            "\n    A variable name should follow the `var` keyword.\n\n");

    if (!tok_matches(par, TOK_SEMICOLON))
        err_at_tok(par->path, par->prev_tok,
            "\n    Missing semicolon after variable declaration.\n\n");

    return stm;
}

static struct stm *
statement(struct par_state *par)
/*  statement -> block_stm
               | if_stm
               | while_stm
               | print_stm
               | expr_stmt                                                   */
{
    struct stm *stm;

    if (tok_matches(par, TOK_DO))
        stm = block(par);
    else if (tok_matches(par, TOK_IF))
        stm = if_cond(par);
    else if (tok_matches(par, TOK_WHILE))
        stm = while_cond(par);
    else if (tok_matches(par, TOK_PRINT))
        stm = print(par);
    else
        stm = expr_stmt(par);

    return stm;
}

static struct stm *
block(struct par_state *par)
/*  block_stm -> ^do^ declaration* "end"                                     */
{
    Vector_T statements;

    statements = Vector_new(sizeof(struct stm *));
    while (!tok_is(par, TOK_END) && !tok_is(par, TOK_EOF))
    {
        struct stm *stm = declaration(par);
        Vector_push(statements, &stm);
    }

    if (!tok_matches(par, TOK_END))
        err_at_tok(par->path, par->prev_tok,
            "\n    Missing `end` keyword after block.\n\n");

    return stm_new_block(statements);
}

static struct stm *
if_cond(struct par_state *par)
/*  if_stm -> ^if^ "(" expression ")" statement ( "else" statement )?        */
{
    struct stm *stm;

    if (tok_matches(par, TOK_PAREN_LEFT))
    {
        struct exp *cond;

        cond = expression(par);

        if (tok_matches(par, TOK_PAREN_RIGHT))
        {
            struct stm *then_block;
            struct stm *else_block;

            then_block = statement(par);

            if (tok_matches(par, TOK_ELSE))
                else_block = statement(par);
            else
                else_block = NULL;

            stm = stm_new_if(cond, then_block, else_block);
        }
        else
            err_at_tok(par->path, par->prev_tok,
                "\n    Expected a closing paren after the if condition.\n\n");
    }
    else
        err_at_tok(par->path, par->prev_tok,
            "\n    Expected an opening paren after `if` keyword.\n\n");

    return stm;
}

static struct stm *
while_cond(struct par_state *par)
/*  while_stm -> ^while^ "(" expression ")" statement                        */
{
    struct stm *stm;

    if (tok_matches(par, TOK_PAREN_LEFT))
    {
        struct exp *cond;

        cond = expression(par);

        if (tok_matches(par, TOK_PAREN_RIGHT))
        {
            struct stm *body;

            body = statement(par);
            stm = stm_new_while(cond, body);
        }
        else
            err_at_tok(par->path, par->prev_tok,
                "\n    Expected a closing paren after the while "
                "condition.\n\n");
    }
    else
        err_at_tok(par->path, par->prev_tok,
            "\n    Expected an opening paren after `while` keyword.\n\n");

    return stm;
}

static struct stm *
print(struct par_state *par)
/*  print_stm -> ^print^ expression ";"                                      */
{
    struct exp *exp;

    exp = expression(par);
    if (!tok_matches(par, TOK_SEMICOLON))
        err_at_tok(par->path, par->prev_tok,
            "\n    Missing semicolon after print statement.\n\n");

    return stm_new_print(exp, par->prev_tok);
}

static struct stm *
expr_stmt(struct par_state *par)
/*  expr_stmt -> expression ";"                                              */
{
    struct exp *exp;

    exp = expression(par);
    if (!tok_matches(par, TOK_SEMICOLON))
        err_at_tok(par->path, par->prev_tok,
            "\n    Missing semicolon after expression statement.\n\n");

    return stm_new_expr_stmt(exp, par->prev_tok);
}

static struct exp *
expression(struct par_state *par)
/*  expression -> assignment                                                 */
{
    return assignment(par);
}

static struct exp *
assignment(struct par_state *par)
/* XXX having assignment be an expression simplifies the grammar, but puts us
   in the situation of having an expression with a side effect. It would be
   suited better to being a statement                                        */
/*  assignment -> equality ( "=" assignment )*                               */
{
    struct exp *left = equality(par);

    if (tok_matches(par, TOK_EQUAL))
    {
        struct tok *eq = par->prev_tok;
        struct exp *exp = assignment(par);

        if (left->type == EXP_VAR)
        {
            struct tok *varname = left->data.name;
            free(left);
            return exp_new_assign(varname, exp);
        }
        else
            err_at_tok(par->path, eq,
                "\n    Invalid assignment target. "
                "Expected a variable name.\n\n");
    }

    return left;
}

static struct exp *
equality(struct par_state *par)
/*  equality -> ordering ( ( "!=" | "==" ) ordering )*                       */
{
    struct exp *left = ordering(par);

    while (tok_matches(par, TOK_BANG_EQUAL) ||
           tok_matches(par, TOK_EQUAL_EQUAL))
    {
        struct tok *op = par->prev_tok;
        struct exp *right = ordering(par);
        left = exp_new_binary(op, left, right);
    }

    return left;
}

static struct exp *
ordering(struct par_state *par)
/*  ordering -> addition ( ( ">" | ">=" | "<" | "<=" ) addition)*          */
{
    struct exp *left = addition(par);

    while (tok_matches(par, TOK_GREAT) || tok_matches(par, TOK_GREAT_EQUAL) ||
           tok_matches(par, TOK_LESS) || tok_matches(par, TOK_LESS_EQUAL))
    {
        struct tok *op = par->prev_tok;
        struct exp *right = addition(par);
        left = exp_new_binary(op, left, right);
    }

    return left;
}

static struct exp *
addition(struct par_state *par)
/*  addition -> multiplication ( ( "-" | "+" ) multiplication)*              */
{
    struct exp *left = multiplication(par);

    while (tok_matches(par, TOK_MINUS) || tok_matches(par, TOK_PLUS))
    {
        struct tok *op = par->prev_tok;
        struct exp *right = multiplication(par);
        left = exp_new_binary(op, left, right);
    }

    return left;
}

static struct exp *
multiplication(struct par_state *par)
/*  multiplication -> unary ( ( "/" | "*" ) unary )*                         */
{
    struct exp *left = unary(par);

    while (tok_matches(par, TOK_SLASH) || tok_matches(par, TOK_STAR))
    {
        struct tok *op = par->prev_tok;
        struct exp *right = unary(par);
        left = exp_new_binary(op, left, right);
    }

    return left;
}

static struct exp *
unary(struct par_state *par)
/*  unary -> ( ("!" | "-") unary )
           | primary                                                         */
{
    if (tok_matches(par, TOK_BANG) || tok_matches(par, TOK_MINUS))
    {
        struct tok *op = par->prev_tok;
        struct exp *exp = unary(par);
        return exp_new_unary(op, exp);
    }
    else
        return primary(par);
}

static struct exp *
primary(struct par_state *par)
/* primary -> IDENTIFIER
            | INTEGER | DOUBLE | STRING | "false" | "true" |
            | "(" expression ")"
            | XXX error productions                                          */
{
    if (tok_matches(par, TOK_IDENTIFIER))
        return exp_new_var(par->prev_tok);
    if (tok_matches(par, TOK_INTEGER) || tok_matches(par, TOK_DOUBLE) ||
        tok_matches(par, TOK_STRING) ||
        tok_matches(par, TOK_FALSE) || tok_matches(par, TOK_TRUE))
    {
        return exp_new_literal(par->prev_tok);
    }
    else if (tok_matches(par, TOK_PAREN_LEFT))
    {
        struct exp *exp = expression(par);
        if (!tok_matches(par, TOK_PAREN_RIGHT))
        {
            err_at_tok(par->path, par->this_tok,
                "\n    Expected a closing paren, instead got `%s`.\n\n",
                par->this_tok->lexeme);
            return NULL; /* Unreachable. */
        }
        else
            return exp;
    }
    else
    {
        err_at_tok(par->path, par->this_tok,
            "\n    Expected number, paren or keyword. Instead got `%s`.\n\n",
            par->this_tok->lexeme);
        return NULL; /* Unreachable. */
    }
}

static struct tok *
tok_next(struct par_state *par)
{
    par->prev_tok = par->this_tok;
    par->this_tok = lex_get_tok(&par->lex);

    return par->prev_tok;
}

static bool
tok_matches(struct par_state *par, enum tok_type type)
{
    if (par->this_tok->type == type)
    {
        tok_next(par);
        return true;
    }
    else
        return false;
}

inline static bool
tok_is(struct par_state *par, enum tok_type type)
{
    return par->this_tok->type == type;
}

inline static bool
tok_was(struct par_state *par, enum tok_type type)
{
    return par->prev_tok->type == type;
}

static struct stm *
stm_new_block(Vector_T block)
{
    struct stm *stm = malloc(sizeof(struct stm));

    stm->type = STM_BLOCK;
    stm->data.block = block;

    return stm;
}

static struct stm *
stm_new_if(struct exp *cond, struct stm *then_block, struct stm *else_block)
{
    struct stm *stm = malloc(sizeof(struct stm));

    stm->type = STM_IF;
    stm->data.if_cond.cond = cond;
    stm->data.if_cond.then_block = then_block;
    stm->data.if_cond.else_block = else_block;

    return stm;
}

struct stm*
stm_new_while(struct exp *cond, struct stm *body)
{
    struct stm *stm = malloc(sizeof(struct stm));

    stm->type = STM_WHILE;
    stm->data.while_cond.cond = cond;
    stm->data.while_cond.body = body;

    return stm;
}

static struct stm *
stm_new_var_decl(struct tok *name, struct exp *exp)
{
    struct stm *stm = malloc(sizeof(struct stm));

    stm->type = STM_VAR_DECL;
    stm->data.var_decl.name = name;
    stm->data.var_decl.exp = exp;

    return stm;
}

static struct stm *
stm_new_expr_stmt(struct exp *exp, struct tok *last)
{
    struct stm *stm = malloc(sizeof(struct stm));

    stm->type = STM_EXPR_STMT;
    stm->data.expr.exp = exp;
    stm->data.expr.last = last;

    return stm;
}

static struct stm *
stm_new_print(struct exp *exp, struct tok *last)
{
    struct stm *stm = malloc(sizeof(struct stm));

    stm->type = STM_PRINT;
    stm->data.print.exp = exp;
    stm->data.print.last = last;

    return stm;
}

static struct exp *
exp_new_assign(struct tok *name, struct exp *value)
{
    struct exp *exp = malloc(sizeof(struct stm));

    exp->type = EXP_ASSIGN;
    exp->data.assign.name = name;
    exp->data.assign.exp = value;

    return exp;
}

static struct exp *
exp_new_binary(struct tok *op, struct exp *left, struct exp *right)
{
    struct exp *exp = malloc(sizeof(struct exp));

    exp->type = EXP_BINARY;
    exp->data.binary.op = op;
    exp->data.binary.left = left;
    exp->data.binary.right = right;

    return exp;
}

static struct exp *
exp_new_unary(struct tok *op, struct exp *operand)
{
    struct exp *exp = malloc(sizeof(struct exp));

    exp->type = EXP_UNARY;
    exp->data.unary.op = op;
    exp->data.unary.exp = operand;

    return exp;
}

static struct exp *
exp_new_group(struct exp *group, struct tok *lparen, struct tok *rparen)
/* XXX group codepaths are neither used nor tested */
{
    struct exp *exp = malloc(sizeof(struct exp));

    exp->type = EXP_GROUP;
    exp->data.group.exp = group;
    exp->data.group.lparen = lparen;
    exp->data.group.rparen = rparen;

    return exp;
}

static struct exp *
exp_new_var(struct tok *tok)
{
    struct exp *exp = malloc(sizeof(struct exp));

    exp->type = EXP_VAR;
    exp->data.name = tok;

    return exp;
}

static struct exp *
exp_new_literal(struct tok *tok)
{
    struct exp *exp = malloc(sizeof(struct exp));

    exp->type = EXP_LITERAL;
    exp->data.literal = tok;

    return exp;
}

void
print_stm(struct stm *stm)
{
    switch (stm->type)
    {
        case STM_BLOCK:
        {
            Vector_T vector;
            int i, len;

            puts("[");
            vector = stm->data.block;
            len = Vector_length(vector);
            for (i = 0; i < len; ++i)
                print_stm(*(struct stm **)Vector_get(vector, i));
            puts("]");
        } break;

        case STM_IF:
        {
            printf("[ if ");
            print_exp(stm->data.if_cond.cond);
            putchar('\n');
            printf("then ");
            print_stm(stm->data.if_cond.then_block);

            if (stm->data.if_cond.else_block != NULL)
            {
                printf("else ");
                print_stm(stm->data.if_cond.else_block);
            }
            puts("]");
        } break;

        case STM_WHILE:
        {
            printf("[ while ");
            print_exp(stm->data.while_cond.cond);
            putchar('\n');
            print_stm(stm->data.while_cond.body);
            puts("]");
        } break;

        case STM_VAR_DECL:
        {
            printf("[ %s = ", stm->data.var_decl.name->lexeme);
            print_exp(stm->data.var_decl.exp);
            puts("]");
        } break;

        case STM_EXPR_STMT:
        {
            printf("[ expstm ");
            print_exp(stm->data.expr.exp);
            puts("]");
        } break;

        case STM_PRINT:
        {
            printf("[ print ");
            print_exp(stm->data.expr.exp);
            puts("]");
        } break;
    }
}

void
print_exp(struct exp *exp)
{
    switch (exp->type)
    {
        case EXP_ASSIGN:
        {
            printf("( assign %s ", exp->data.assign.name->lexeme);
            print_exp(exp->data.assign.exp);
            printf(") ");
        } break;

        case EXP_BINARY:
        {
            printf("( %s ", exp->data.binary.op->lexeme);
            print_exp(exp->data.binary.left);
            print_exp(exp->data.binary.right);
            printf(") ");
        } break;

        case EXP_UNARY:
        {
            printf("( %s ", exp->data.unary.op->lexeme);
            print_exp(exp->data.unary.exp);
            printf(") ");
        } break;

        case EXP_GROUP:
        {
            printf("( ");
            print_exp(exp->data.group.exp);
            printf(") ");
        } break;

        case EXP_VAR:
            printf("%s ", exp->data.name->lexeme);
        break;

        case EXP_LITERAL:
        {
            struct tok *tok = exp->data.literal;
            if (tok->type == TOK_STRING)
                printf("\"%s\" ", exp->data.literal->lexeme);
            else
                printf("%s ", exp->data.literal->lexeme);
        } break;
    }
}

File: src/pymidas/parser.py
from lexer import Token, Lexer
from toks import *
from stmt import *
from expr import *


class ParserError(Exception):
    pass


class Parser():
    def __init__(self):
        self.lex = Lexer()
        self.path = None
        self.prev_tok = None
        self.this_tok = None
        self.loop_depth = 0
        self.fun_depth = 0

    def parse(self, path):
        with open(path, "r") as source:
            text = source.read() + "\0"
        self.lex.feed(text)
        self.tok_next()
        return self.program()

    def program(self):
        # program -> statement* "EOF"
        stmts = []
        while not self.tok_is(TOK_EOF):
            stm = self.declaration()
            stmts.append(stm)
        return stmts

    def declaration(self):
        # declaration -> var_decl ";"
        #              | fun_decl ";"
        #              | statement
        if self.tok_matches(TOK_VAR):
            stm = self.var_decl()
        elif self.tok_matches(TOK_FUN):
            stm = self.fun_decl()
        else:
            stm = self.statement()
        return stm

    def var_decl(self):
        # var_decl -> ^var^ identifier "=" expression ";"
        self.tok_consume(TOK_IDENTIFIER,
                "A variable should follow the `var` keyword")
        name = self.prev_tok.lexeme
        self.tok_consume(TOK_EQUAL,
                "An equal sign should follow the variable's name")
        value = self.expression()
        self.tok_consume(TOK_SEMICOLON,
                "Missing semicolon after variable declaration")
        return StmVarDecl(name, value)

    def fun_decl(self):
        # fun_decl -> ^fun^ identifier "(" formals? ")" block
        # formals  -> identifier ( "," identifier )*
        self.tok_consume(TOK_IDENTIFIER,
                "A function name should follow the `fun` keuyword")
        name = self.prev_tok.lexeme
        self.tok_consume(TOK_PAREN_LEFT,
                "Expected an opening paren after the function name")
        formals = []

        if not self.tok_matches(TOK_PAREN_RIGHT):
            self.tok_consume(TOK_IDENTIFIER,
                    "Function parameters should be a comma-separated list" +
                    "of identifiers")
            formals.append(self.prev_tok.lexeme)

            while self.tok_matches(TOK_COMMA):
                self.tok_consume(TOK_IDENTIFIER,
                        "Function parameters should be a comma-separated " +
                        "list of identifiers")
                formals.append(self.prev_tok.lexeme)
            self.tok_consume(TOK_PAREN_RIGHT,
                    "Expected a closing paren after function arguments")
        self.tok_consume(TOK_DO,
                "Expected a do...end block after function parameter list")
        self.fun_depth += 1
        body = self.block()
        self.fun_depth -= 1

        return StmFunDecl(name, formals, body)

    def statement(self):
        # statement -> block
        #            | if_stm
        #            | while_stm
        #            | for_stm
        #            | break_stm
        #            | return_stm
        #            | print_stm
        #            | exp_stm
        if self.tok_matches(TOK_DO):
            stm = self.block()
        elif self.tok_matches(TOK_IF):
            stm = self.if_cond()
        elif self.tok_matches(TOK_WHILE):
            stm = self.while_cond()
        elif self.tok_matches(TOK_FOR):
            stm = self.for_cond()
        elif self.tok_matches(TOK_BREAK):
            stm = self.break_stm()
        elif self.tok_matches(TOK_RETURN):
            stm = self.return_stm()
        elif self.tok_matches(TOK_PRINT):
            stm = self.print_stm()
        else:
            stm = self.exp_stm()

        return stm

    def block(self):
        # block -> ^do^ declaration* "end"
        stmts = []
        while not self.tok_is(TOK_END) and not self.tok_is(TOK_EOF):
            stm = self.declaration()
            stmts.append(stm)
        self.tok_consume(TOK_END,
                "Missing `end` keyword after block")
        return StmBlock(stmts)

    def if_cond(self):
        # if_stm -> ^if^ "(" expression ")" statement ( "else" statement )?
        self.tok_consume(TOK_PAREN_LEFT,
                "Expected an opening paren after `if` keyword")
        cond = self.expression()
        self.tok_consume(TOK_PAREN_RIGHT,
                "Expected a closing paren after the if condition")
        then_block = self.statement()

        if self.tok_matches(TOK_ELSE):
            else_block = self.statement()
        else:
            else_block = None
        return StmIf(cond, then_block, else_block)

    def while_cond(self):
        # while_stm -> ^while^ "(" expresion ")" statement
        self.tok_consume(TOK_PAREN_LEFT,
                "Expected an opening paren after `while` keyword")
        cond = self.expression()
        self.tok_consume(TOK_PAREN_RIGHT,
                "Expected a closing paren after the while condition")
        self.loop_depth += 1
        body = self.statement()
        self.loop_depth -= 1

        return StmWhile(cond, body)

    def for_cond(self):
        # for_stm -> ^for^ "("
        #                      ( var_decl | exp_stm | ";" )
        #                      expression? ";"
        #                      assignment?
        #                  ")" statement
        # NB: var_decl and exp_stm already contain a semicolon
        self.tok_consume(TOK_PAREN_LEFT,
                "Expected an opening paren after the `for` keyword")
        if self.tok_matches(TOK_VAR):
            init = self.var_decl()
        elif self.tok_matches(TOK_SEMICOLON):
            init = None
        else:
            init = self.exp_stm()

        if not self.tok_matches(TOK_SEMICOLON):
            cond = self.expression()
            self.tok_consume(TOK_SEMICOLON,
                    "Expected a semicolon after the for condition")
        else:
            cond = None

        if not self.tok_matches(TOK_PAREN_RIGHT):
            incr = self.assignment()
            self.tok_consume(TOK_PAREN_RIGHT,
                    "Expected a closing paren after the for construct")
        else:
            incr = None

        self.loop_depth += 1
        body = self.statement()
        self.loop_depth -= 1

        if incr != None:
            incr_stm = StmExp(incr)
            if body.kind == STM_BLOCK:
                body.block.append(incr_stm)
            else:
                stmts = []
                stmts.append(body)
                stmts.append(incr_stm)
                body = StmBlock(stmts)

        if cond == None:
            tok = Token("true", TOK_TRUE, 4, 0, 0)
            cond = ExpLiteral(tok)

        loop = StmWhile(cond, body)

        if init != None:
            stmts = []
            stmts.append(init)
            stmts.append(loop)
            loop = StmBlock(stmts)

        return loop

    def break_stm(self):
        # ^break^ ";"
        if self.loop_depth != 0:
            self.tok_consume(TOK_SEMICOLON,
                    "Missing semicolon after break statement")
        else:
            raise ParserError("Encountered break statement outside of a for " +
                              "or while loop")
        return StmBreak()

    def return_stm(self):
        # return_stm -> ^return^ expression? ";"
        if self.fun_depth != 0:
            if not self.tok_matches(TOK_SEMICOLON):
                ret_exp = self.expression()
                self.tok_consume(TOK_SEMICOLON,
                        "Missing semicolon after return statement")
                return StmReturn(ret_exp)
            else:
                nil = Token("nil", TOK_NIL, 3, 0, 0)
                return StmReturn(ExpLiteral(nil))
        else:
            raise ParserError("Encountered a return statements outside of a " +
                    "function declaration")

    def print_stm(self):
        # print_stm -> ^print^ expression ";"
        exp = self.expression()
        self.tok_consume(TOK_SEMICOLON, "Missing semicolon after print statement")
        return StmPrint(exp)

    def exp_stm(self):
        # exp_stm -> expression ";"
        exp = self.expression()
        self.tok_consume(TOK_SEMICOLON,
                "Missing semicolon after expression statement")
        return StmExp(exp)

    def expression(self):
        # expression -> assigment
        return self.assignment()

    def assignment(self):
        # assignment -> concat ( "=" assignment )
        left = self.concat()
        if self.tok_matches(TOK_EQUAL):
            right = self.assignment()
            if left.kind == EXP_IDENT:
                return ExpAssign(left.name, right)
            else:
                raise ParserError("Cannot assign to this target")
        return left

    def concat(self):
        # concat -> or ( "++" or )*
        left = self.logic_or()
        while self.tok_matches(TOK_PLUS_PLUS):
            op = self.prev_tok.kind
            right = self.logic_or()
            left = ExpBinary(op, left, right)
        return left

    def logic_or(self):
        # or -> and ( "or" and )*
        left = self.logic_and()
        while self.tok_matches(TOK_OR):
            op = self.prev_tok.kind
            right = self.logic_and()
            left = ExpBinary(op, left, right)
        return left

    def logic_and(self):
        # and -> equality ( "or" equality )*
        left = self.equality()
        while self.tok_matches(TOK_AND):
            op =  self.prev_tok.kind
            right = self.equality()
            left = ExpBinary(op, left, right)
        return left

    def equality(self):
        # equality -> ordering ( ( "!=" | "==" ) ordering )*
        left = self.ordering()
        while self.tok_matches(TOK_BANG_EQUAL) or \
                self.tok_matches(TOK_EQUAL_EQUAL):
            op = self.prev_tok.kind
            right = self.ordering()
            left = ExpBinary(op, left, right)
        return left

    def ordering(self):
        # ordering -> addition ( ( ">" | ">=" | "<" | "<=" ) addition )*
        left = self.addition()
        while self.tok_matches(TOK_GREAT) or \
                self.tok_matches(TOK_GREAT_EQUAL) or \
                self.tok_matches(TOK_LESS) or \
                self.tok_matches(TOK_LESS_EQUAL):
            op = self.prev_tok.kind
            right = self.addition()
            left = ExpBinary(op, left, right)
        return left

    def addition(self):
        # addition -> multiplication ( ( "-" | "+" ) multiplication )*
        left = self.multiplication()
        while self.tok_matches(TOK_MINUS) or \
                self.tok_matches(TOK_PLUS):
            op = self.prev_tok.kind
            right = self.multiplication()
            left = ExpBinary(op, left, right)
        return left

    def multiplication(self):
        # multiplication -> unary ( ( "/" | "//" | "*" | "%" ) unary )*
        left = self.unary()
        while self.tok_matches(TOK_SLASH) or \
                self.tok_matches(TOK_SLASH_SLASH) or \
                self.tok_matches(TOK_STAR) or \
                self.tok_matches(TOK_PERCENT):
            op = self.prev_tok.kind
            right = self.unary()
            left = ExpBinary(op, left, right)
        return left

    def unary(self):
        # unary -> ( ( "!" | "-" ) unary )
        #        | call
        if self.tok_matches(TOK_BANG) or self.tok_matches(TOK_MINUS):
            op = self.prev_tok.kind
            exp = self.unary()
            return ExpUnary(op, exp)
        else:
            return self.call()

    def call(self):
        # call -> primary ( "(" parameters? ")" )*
        callee = self.primary()
        while self.tok_matches(TOK_PAREN_LEFT):
            params = []
            if not self.tok_matches(TOK_PAREN_RIGHT):
                exp = self.expression()
                params.append(exp)
                while self.tok_matches(TOK_COMMA):
                    exp = self.expression()
                    params.append(exp)
                self.tok_consume(TOK_PAREN_RIGHT,
                        "Expected a closing paren after arguments")
            callee = ExpCall(callee, params)
        return callee

    def primary(self):
        # primary -> IDENTIFIER
        #          | INTEGER | DOUBLE | STRING | NIL | "false" | "true" |
        #          | "(" expression ")"
        if self.tok_matches(TOK_IDENTIFIER):
            exp = ExpIdent(self.prev_tok.lexeme)
        elif self.tok_matches(TOK_INTEGER) or \
                self.tok_matches(TOK_DOUBLE) or \
                self.tok_matches(TOK_STRING) or \
                self.tok_matches(TOK_NIL) or \
                self.tok_matches(TOK_FALSE) or \
                self.tok_matches(TOK_TRUE):
            exp = ExpLiteral(self.prev_tok)
        elif self.tok_matches(TOK_PAREN_LEFT):
            exp = self.expression()
            self.tok_consume(TOK_PAREN_RIGHT, "Expected a closing paren")
        else:
            raise ParserError("Expected number, paren or keyword. Got " +
                    str(self.this_tok))
        return exp

    def tok_next(self):
        self.prev_tok = self.this_tok
        self.this_tok = self.lex.get_tok()
        return self.prev_tok

    def tok_matches(self, kind):
        if self.this_tok.kind == kind:
            self.tok_next()
            return True
        else:
            return False

    def tok_consume(self, kind, message):
        if self.this_tok.kind == kind:
            self.tok_next()
        else:
            raise ParserError(message)

    def tok_is(self, kind):
        return self.this_tok.kind == kind

    def tok_was(self, kind):
        return self.prev_tok.kind == kind
